pkg/api/registry/registry.proto

```diff
@@ -22,61 +22,48 @@ message Destination {
     uint32 weight = 2;
 }
 
-message NetworkServiceManager {
-    string name = 1;
-    string url = 2;
-    google.protobuf.Timestamp expiration_time = 3;
-    string state = 4;
+message NetworkServiceLabels {
+    map<string, string> labels = 1;
 }
 
 message NetworkServiceEndpoint {
     string name = 1;
-    string payload = 2;
-    string network_service_name = 3;
-    string network_service_manager_name = 4;
-    map<string, string> labels = 5;
-    string state = 6;
-
+    repeated string network_service_name = 2;
+    map<string, NetworkServiceLabels> network_service_labels = 3;
     // Endpoint callback address, could be any of unix/tcp/callback schema ones.
-    string url = 7;
+    string url = 4;
+    // Network service manager URL for the endpoint.
+    string network_service_manger_url = 5;
+    google.protobuf.Timestamp expiration_time = 6;
 }
 
-message FindNetworkServiceRequest {
-    string network_service_name = 1;
+service NetworkServiceEndpointRegistry {
+    rpc Register (NetworkServiceEndpoint) returns (NetworkServiceEndpoint);
+    rpc Monitor (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpoint);
+    rpc Find (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpointList);
+    rpc Unregister (NetworkServiceEndpoint) returns (google.protobuf.Empty);
 }
 
-message FindNetworkServiceResponse {
-    string payload = 1;
-    NetworkService network_service = 2;
-    map<string, NetworkServiceManager> network_service_managers = 3;
-    repeated NetworkServiceEndpoint network_service_endpoints = 4;
+service NetworkServiceRegistry {
+    rpc Register (NetworkService) returns (NetworkService);
+    rpc Monitor (NetworkServiceQuery) returns (stream NetworkService);
+    rpc Find (NetworkServiceQuery) returns (stream NetworkServiceList);
+    rpc Unregister (NetworkService) returns (google.protobuf.Empty);
 }
 
-message NSERegistration {
+message NetworkServiceQuery {
     NetworkService network_service = 1;
-    NetworkServiceManager network_service_manager = 2;
-    NetworkServiceEndpoint network_service_endpoint = 3;
-}
-
-message RemoveNSERequest {
-    string network_service_endpoint_name = 1;
 }
 
-service NetworkServiceRegistry {
-    rpc RegisterNSE (NSERegistration) returns (NSERegistration);
-    rpc BulkRegisterNSE (stream NSERegistration) returns (stream NSERegistration);
-    rpc RemoveNSE (RemoveNSERequest) returns (google.protobuf.Empty);
+message NetworkServiceEndpointQuery {
+    NetworkServiceEndpoint network_service_endpoint = 1;
+    NetworkService network_service = 2;
```

Comment:
Why do we have both a NetworkServiceEndpoint and a NetworkService object in this query?

Reply:
Related to this comment https://github.com/networkservicemesh/api/pull/41#issuecomment-638912388

Reply:
The QPS for gRPC in go is pretty high.

https://performance-dot-grpc-testing.appspot.com/explore?dashboard=5636470266134528&widget=828849126&container=1808124444

I'd prefer to start without a list and introduce a list if gRPC throughput turns out to be our bottleneck when we perform stress tests.

We also have other techniques we can use to reduce the total quantity of endpoints transferred such as summarization.

Reply:
@fkautz I've little tested locally(If you are interested I can provide the testing on CI) and got next results:


n = 1e7

Steam of lists where the length of the list is 10000, time :11.556816ms
Steam of single object: time: 30.680219342s

n = 1e9

Steam of lists where the length of the list is 10000, time :14.89991ms
Steam of single object: time: 30.680219342s


n = 1e10

Steam of lists where the length of the list is 10000, time: 46.025932ms
Steam of single object: time: 5m10.775024696s


_n is count of endpoinst.



Also, I've tested the case when used list without stream. It works less than second only with `n=6`, time: 621.380154ms (for n=7 time is 6.172965725s).

Reply:
So it looks like a stream of list has the best performance.

Reply:
> So it looks like a stream + list has the best performance.

True, and interesting... but perf is pretty close up to 10^7... I'm also most curious the *why* under the increasing perf difference... it feels like there is something interesting in the system going on.

I tend to agree with @fkautz , that we should start with the non-list stream case.  I'm pretty sure that by the time you are playing in the 10^7 range, other elements of the system, like the underlying DB being read from will be the dominant factor.

Reply:
My preference is still for a single object rather than a list for simplicity.

In the case of n = 1e7:

Steam of lists where the length of the list is 10000, time :11.556816ms
Steam of single object: time: 22.952515ms

That's a theoretical limit of approximately 450 million messages / second. There are articles from companies like Netflix about how they achieved 1 million inserts / second in a 60 node Cassandra cluster. I think the single object approach is ok in this scenario. It's highly unlikely gRPC will be our bottleneck here. If it is, we can horizontally scale it. Even if you go with an append only queue such as Kafka, you're looking at 1 or 2M appends / second for a cluster of 3.

In short, gRPC will not be our bottleneck. We're looking at database and KV stores as the bottleneck here.

Reply:
That being said, I do worry about latency between systems. It would be good to see what happens when you add 10ms or 100ms of latency to the network. That may provide better context on how we proceed.

Reply:
> That being said, I do worry about latency between systems. It would be good to see what happens when you add 10ms or 100ms of latency to the network. That may provide better context on how we proceed.

I very like the direction you are thinking here, I think we need to add this kind of stress tests when main NSM components will be completed to see what in the system is our bottleneck.